### PR TITLE
Add telemetry to dogstatsd to better troubleshoot performance

### PR DESCRIPTION
**Pull request by hush-hush**
### What does this PR do?

We now track how long a listeners is not reading data and how long
channel communication takes in dogstatsd.

### Describe your test plan

Enables the agent telemetry and curl ` http://localhost:5000/telemetry` to see the new metrics.

### Performance

We deploy this PR vs master on our benchmark environment without any visible overhead from the new prometheus histogram

### CODE REVIEW

File: pkg/telemetry/histogram.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+// Histogram tracks the value of one health metric of the Agent.
+type Histogram interface {
+	// Observe  the value to the Gauge value.
+	Observe(value float64, tagsValue ...string)
+	// Delete deletes the value for the Gauge with the given tags.
+	Delete(tagsValue ...string)
+}
+
+// NewGauge creates a Gauge with default options for telemetry purpose.
+// Current implementation used: Prometheus Gauge
```
Comment (olivielpeau) [inline pkg/telemetry/histogram.go:23]: comment needs an update (applies to a lot of other comments in your changes to `telemetry` that still refer to Gauge)

File: pkg/dogstatsd/listeners/telemetry.go [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package listeners
+
+import "github.com/DataDog/datadog-agent/pkg/telemetry"
+
+var (
+	tlmListener = telemetry.NewHistogram("dogstatsd", "listener_read_latency",
+		[]string{"listener_type"}, "Time in nanoseconds while the listeners is not reading data", []float64{100, 200, 300, 500, 1000})
```
Comment (olivielpeau) [inline pkg/dogstatsd/listeners/telemetry.go:7]:
```suggestion
		[]string{"listener_type"}, "Time in nanoseconds while the listener is not reading data", []float64{100, 200, 300, 500, 1000})
```

File: pkg/dogstatsd/listeners/udp.go [hunk 3]
```diff
@@ -93,10 +94,13 @@ func NewUDPListener(packetOut chan Packets, sharedPacketPool *PacketPool) (*UDPL
 
 // Listen runs the intake loop. Should be called in its own goroutine
 func (l *UDPListener) Listen() {
+	var t1, t2 time.Time
 	log.Infof("dogstatsd-udp: starting to listen on %s", l.conn.LocalAddr())
 	for {
 		udpPackets.Add(1)
```
Comment (olivielpeau) [inline pkg/dogstatsd/listeners/udp.go:100]: unrelated to your changes, but this line should be moved to after the `ReadFrom` call, next to `tlmUDPPackets.Inc("ok")`, probably good to fix this while you're at it.

**Review by olivielpeau — COMMENTED** [review]
This looks good at a high level! Left a few comments inline

File: pkg/config/config.go [hunk 4]
```diff
@@ -721,6 +721,14 @@ func InitConfig(config Config) {
 	// This create a lot of billable custom metrics.
 	config.BindEnvAndSetDefault("telemetry.enabled", false)
 	config.SetKnown("telemetry.checks")
+	// We're using []string as a default instead of []float64 because viper can only parse list of string from the environment
+	//
+	// The histogram buckets use to track the time in nanoseconds DogStatsD listeners are not reading/waiting new data
+	config.BindEnvAndSetDefault("telemetry.dogstatsd_listeners_latency_buckets", []string{})
```
Comment (truthbk) [inline pkg/config/config.go:727]: should we have a subsection for dogstatsd telemetry: `telemetry.dogstatsd.*` ?

**Review by truthbk — COMMENTED** [review]
Looks great, I think my only concern is that the current implementation doesn't allow to disable the telemetry (unless I missed something), the telemetry listeners always have the arithmetic computation taking place when we call `Observe()` so the (small) overhead is always there. Do we want to consider making the telemetry optional?

**Review by truthbk — APPROVED** [review]
Thanks for validating the negligible performance overhead! 🙇

**Pull request merged**